WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/GenerateProductWithAI.kt

```
@@ -0,0 +1,129 @@
+import com.google.gson.annotations.SerializedName
+import com.woocommerce.android.ai.AIRepository
+import com.woocommerce.android.extensions.isNotNullOrEmpty
+import com.woocommerce.android.model.Product
+import com.woocommerce.android.model.ProductCategory
+import com.woocommerce.android.model.ProductTag
+import com.woocommerce.android.ui.products.ParameterRepository
+import com.woocommerce.android.ui.products.ProductHelper
+import com.woocommerce.android.ui.products.ProductStatus.DRAFT
+import com.woocommerce.android.ui.products.ProductType.SIMPLE
+import com.woocommerce.android.ui.products.ai.AboutProductSubViewModel.AiTone
+import com.woocommerce.android.ui.products.categories.ProductCategoriesRepository
+import com.woocommerce.android.ui.products.models.SiteParameters
+import com.woocommerce.android.ui.products.tags.ProductTagsRepository
+import com.woocommerce.android.util.WooLog
+import com.woocommerce.android.util.WooLog.T.AI
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.math.BigDecimal
+import javax.inject.Inject
+
+class GenerateProductWithAI @Inject constructor(
+    private val aiRepository: AIRepository,
+    private val categoriesRepository: ProductCategoriesRepository,
+    private val tagsRepository: ProductTagsRepository,
+    private val parametersRepository: ParameterRepository
+) {
+    @Suppress("ReturnCount")
+    suspend operator fun invoke(
+        productName: String,
+        productKeyWords: String,
+        tone: AiTone,
+        languageISOCode: String
+    ): Result<Product> {
+        val existingCategories = categoriesRepository.fetchProductCategories().getOrElse {
+            WooLog.e(AI, "Failed to fetch product categories", it)
+            return Result.failure(it)
+        }
+        val existingTags = tagsRepository.fetchProductTags().getOrElse {
+            WooLog.e(AI, "Failed to fetch product tags", it)
+            return Result.failure(it)
+        }
```

Hmm, from a UX perspective I'm unsure if its nice to block the user from using the whole product generation just because we failed to update the existing tags or categories. Would it make more sense to simply remove the categories and tags from the AI prompt if we fail to fetch them here, while keeping the rest of the feature functional? Wdyt?

---

Good point, I implemented this way because this what we agreed on p1696341676436929/1696330507.148209-slack-C03L1NF1EA3 🙂 

> Would it make more sense to simply remove the categories and tags from the AI prompt if we fail to fetch them here, while keeping the rest of the feature functional? Wdyt?

That's a good alternative, but it won't be straight forward to implement, as we'll need a third version of the prompt, and we need a way to pass the information that fetching failed to generation code, IMO we should wait and see how frequent the failures are, as IMO they should be quite rare, the sites we work here should be more performant than most sites.
cc @selanthiraiyan, WDYT about the above idea?

---

Yeah, I agree that measuring how often this happens is better before investing more efforts in something potentially being very corner case.

---

Great, I'll proceed with merging the PR, and we can always adjust if needed later.

---

> it won't be straight forward to implement, as we'll need a third version of the prompt, and we need a way to pass the information that fetching failed to generation code,

I agree. It won't be straightforward in iOS either. 

Instead of adding a new variation of the prompt, we could also retry fetching the categories/tags before saving the product as a draft. 

Flow,
- Fetching categories/tags fails. 
- We ignore the error and pass an empty array of categories/tags, making AI generate new items.
- We show the new categories/tags in Preview. 
- Before saving as a draft, retry fetching the categories/tags.
- Match the categories/tags and update the local product.
- Proceed to save the product as a draft. 

WDYT?

---

I agree that we should keep an eye on the metrics and make this improvement only if needed.

---

> Instead of adding a new variation of the prompt, we could also retry fetching the categories/tags before saving the product as a draft.

Good idea, the only downside I see is that it would lead to duplicating existing categories when the AI suggests them with small differences, just like the scenario you explained here p1696341612045409/1696330507.148209-slack-C03L1NF1EA3.
If we want to reduce the chances of this, we should use the list of cached categories if it exists, so instead of passing an empty list, we pass the cached list.

---

Good point. I agree that we can send the cached list. 👍